fog/fog_timestamps/src/watcher.rs

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2018-2021 The MobileCoin Foundation
+
+use mc_common::logger::{log, Logger};
+use mc_transaction_core::BlockIndex;
+use mc_watcher::watcher_db::WatcherDB;
+use mc_watcher_api::TimestampResultCode;
+use std::time::{Duration, Instant};
+
+/// Poll for new data every 10 ms
+const POLLING_FREQUENCY: Duration = Duration::from_millis(10);
+/// If a database invariant is violated, e.g. we get block but not block
+/// contents, it typically will not be fixed and so we won't be able to
+/// proceed. But bringing the server down is costly from ops POV because
+/// we will lose all the user rng's.
+///
+/// So instead, if this happens, we log an error, and retry in 1s.
+/// This avoids logging at > 1hz when there is this error, which would be
+/// very spammy. But the retries are unlikely to eventually lead to
+/// progress. Another strategy might be for the server to enter a
+/// "paused" state and signal for intervention.
+const ERROR_RETRY_FREQUENCY: Duration = Duration::from_millis(1000);
+
+// Get the timestamp from the watcher, or an error code,
+// using retries if the watcher fell behind
+pub fn get_watcher_timestamp(
+    block_index: BlockIndex,
+    watcher: &WatcherDB,
+    watcher_timeout: Duration,
+    logger: &Logger,
+) -> u64 {
+    // Timer that tracks how long we have had WatcherBehind error for,
+    // if this exceeds watcher_timeout, we log a warning.
+    let mut watcher_behind_timer = Instant::now();
+    loop {
+        match watcher.get_block_timestamp(block_index) {
+            Ok((ts, res)) => match res {
+                TimestampResultCode::WatcherBehind => {
+                    if watcher_behind_timer.elapsed() > watcher_timeout {
+                        log::warn!(logger, "watcher is still behind on block index = {} after waiting {} seconds, ingest will be blocked", block_index, watcher_timeout.as_secs());
```

Comment:
`ingest` no longer applies here. Applies to other places below.

Reply:
please let me know if below is the change you are requesting: 

log::warn!(logger, "watcher is still behind on block index = {} after waiting {} seconds, caller will be blocked", block_index, watcher_timeout.as_secs());

Reply:
> Any reason to not have this in the `mobilecoin` repository? This is not fog-specific.
> Since this heavily relies on `mc-watcher`, maybe it should live inside that crate?

Ping on this.